### PR TITLE
Implement Deref<Target=T> for JS<T> where T: Reflectable

### PR DESCRIPTION
**Pull request by nox**
We can only borrow `JS<T>` from rooted things, so it's safe to deref it.
The only types that provide mutable `JS<T>` things are `MutHeap<JS<T>>` and
`MutNullableHeap<JS<T>>`, which don't actually expose that they contain
`JS<T>` values.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8060)

### CODE REVIEW

**Comment by eefriedman** [comment]
> The only types that provide mutable JS<T> things are `MutHeap<JS<T>>` and
> `MutNullableHeap<JS<T>>`

That isn't completely true.  For example, https://github.com/servo/servo/blob/e31ad011033ae076efecf61888daa1f6a5900589/components/script/script_task.rs#L409 (`mouse_over_targets: DOMRefCell<Vec<JS<Node>>>,`).

I'd be much more comfortable with this if we unified `JS<T>` and `Root<T>` along the lines of http://manishearth.github.io/blog/2015/09/01/designing-a-gc-in-rust/ .

**Comment by nox** [comment]
The case you point out isn't unsafe in terms of Rust, in the sense that only a panic will happen if we try to drop it while borrowed, AFAIK.

**Comment by eefriedman** [comment]
The way Deref<T> makes the problem worse is that if you write something like `mouse_over_targets.borrow_mut().pop().unwrap()`, you can access the contents without realizing you forgot to root it.  Might not be worth worrying about?

**Comment by nox** [comment]
Unwrapping the result should make @Manishearth's lint complain, because it puts JS<T> on the stack.

**Comment by eefriedman** [comment]
Yes, the lint is probably good enough that it isn't a big deal, just a minor thing to try and watch for when reviewing DOM code.

**Comment by Manishearth** [comment]
We should perhaps disallow cloning `JS<T>`?

**Comment by nox** [comment]
@Manishearth I thought we had removed that with `Copy`, will look into it.

**Comment by Manishearth** [comment]
Isn't `Copy` worse?

**Comment by nox** [comment]
That's what I'm saying, I thought we had removed `Clone` at the same time we removed `Copy`.

**Comment by Manishearth** [comment]
Still there: https://github.com/servo/servo/pull/8060/files#diff-77e42f966279a10a63070b5c62117a56R143

**Comment by nox** [comment]
We can't remove `Clone`, `Sink` heavily rely on it. But I don't think it's a problem, for the same reason as @eefriedman's example isn't.

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -549,7 +549,6 @@ impl Document {
     pub fn send_title_to_compositor(&self) {
         let window = self.window();
         // FIXME(https://github.com/rust-lang/rust/issues/23338)
-        let window = window.r();
```
Comment (dzbarsky) [inline components/script/dom/document.rs]: let compositor = self.window().compositor() and remove the above line and FIXME?

Comment (nox) [inline components/script/dom/document.rs]: Good catch.

**Comment by Ms2ger** [comment]
@bors-servo r+
-S-awaiting-review +S-awaiting-merge 


---

Reviewed 23 of 23 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8060)

**Pull request merged**